src/uopool/services/sanity_check.rs

```diff
@@ -0,0 +1,559 @@
+    ) -> Result<(), BadUserOperationError<M>> {
+        if !user_operation.paymaster_and_data.is_empty() {
+            let paymaster_address = if user_operation.paymaster_and_data.len() >= 20 {
+                Address::from_slice(&user_operation.paymaster_and_data[0..20])
+            } else {
+                return Err(BadUserOperationError::PaymasterVerification {
+                    paymaster_and_data: user_operation.paymaster_and_data.clone(),
+                });
+            };
+
+            let mempool_id = mempool_id(entry_point, &self.chain_id);
+
+            if let Some(entry_point) = self.entry_points.get(&mempool_id) {
+                let deposit_info = entry_point
+                    .get_deposit_info(&paymaster_address)
+                    .await
+                    .map_err(|_| BadUserOperationError::PaymasterVerification {
+                        paymaster_and_data: user_operation.paymaster_and_data.clone(),
+                    })?;
+
+                if U256::from(deposit_info.deposit) > user_operation.max_fee_per_gas {
+                    if let Some(reputation) = self.reputations.read().get(&mempool_id) {
+                        if reputation.get_status(&paymaster_address) != ReputationStatus::BANNED {
+                            self.sanity_check_results
+                                .write()
+                                .entry(user_operation.hash(&entry_point.address(), &self.chain_id))
+                                .or_insert_with(Default::default)
+                                .insert(SanityCheckResult::PaymasterVerified);
+                        }
+                    }
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    fn call_gas_limit(
+        &self,
+        user_operation: &UserOperation,
+    ) -> Result<(), BadUserOperationError<M>> {
+        let non_zero_value_call = gas::non_zero_value_call();
+        if user_operation.call_gas_limit < non_zero_value_call {
+            return Err(BadUserOperationError::LowCallGasLimit {
+                call_gas_limit: user_operation.call_gas_limit,
+                non_zero_value_call,
+            });
+        }
+
+        Ok(())
+    }
+
+    async fn max_fee_per_gas(
+        &self,
+        user_operation: &UserOperation,
+    ) -> Result<(), BadUserOperationError<M>> {
+        let base_fee_estimation = self
+            .gas_oracle
+            .fetch()
+            .await
+            .map_err(|error| BadUserOperationError::GasOracleError(error))?;
+
+        if user_operation.max_priority_fee_per_gas > user_operation.max_fee_per_gas {
+            return Err(BadUserOperationError::HighMaxPriorityFeePerGas {
+                max_priority_fee_per_gas: user_operation.max_priority_fee_per_gas,
+                max_fee_per_gas: user_operation.max_fee_per_gas,
+            });
+        }
+
+        if base_fee_estimation + user_operation.max_priority_fee_per_gas
+            > user_operation.max_fee_per_gas
```

Comment:
I don't understand the calculation here. Could you elaborate a little bit?

I thought it would be just
```rust
max_fee_per_gas > config.minimum_gas && max_fee_per_gas > base_fee
max_priority_fee_per_gas > config.minimum_gas && max_priority_fee_per_gas > base_fee
```
I don't quite understand the logic here.

Reply:
What do you mean here by config.minimum_gas here? Bundler will decide if user operations pay enough fees solely on the current base fee in the network (if it makes sense to include this user operation based on the existing network conditions).

The Max priority fee per gas is the tip the user wants to pay the bundler to include his user operation. The sum of the base fee and tip should be bigger than the max fee per gas (ma fee is the maximum user will pay for user operation).

Your verification would not work because max priority is a much smaller number than the base fee (it's just a tip, it could be only 1 wei in theory: https://docs.alchemy.com/docs/maxpriorityfeepergas-vs-maxfeepergas).

Here we only check the combination of values that don't make sense, and that is if the max priority is bigger than the max fee and if the sum of the base fee and max priority is bigger than the max fee (which means the user operation is offering too low fee to be included).

Reply:
`minimum_gas` means the minimum gas  the bundler want at least which based on the bundler. I am not sure whether I understand the spec correctly here but apparently, we have different understandings.

Let me revisit the spec again.

Reply:
Hmm, I checked the spec again. Maybe we are both right.

`The maxFeePerGas and maxPriorityFeePerGas are above a configurable minimum value that the client is willing to accept. At the minimum, they are sufficiently high to be included with the current block.basefee.`

The first sentence is what you meant, you're right, I missed that. But my understanding of the second sentence is the calculation in the code. So max fee is what use wants to pay maximum with base fee being burned, and max priority being a tip (so base fee + tip must be less or equal to max fee).

Reply:
Cool. I get it now. We are both right:)

Reply:
For now, I added only configurable min maxPriorityFeePerGas since it seems more intuitive for me to have it configurable. I think we can add configurable min maxFeePerGas later.